### DISCUSSION TITLE
[Question] Enabling Extra axes, buttons and (possibly) POVs?

### DISCUSSION THREAD

**Discussion by joao678** [samuelgr/Xidi#79] [Q&A] [answered]
Since Xidi was designed around the xbox controller (or xinput in this case), and I was working around removing this limitation by allowing external programs to feed data to it [(here)](https://github.com/joao678/Xidi/)
And I've ran into some problems...

In `ControllerTypes.h` in the `EButton` enum there are 16 possible buttons but only 12 buttons can be possibly used in mappers one for each xbox button.
By modifying the number of buttons in both `ControllerTypes.h`

https://github.com/samuelgr/Xidi/blob/dd84968cb4a97c4123a29d1f4899b0f9858a7adb/Include/Xidi/Internal/ControllerTypes.h#L91-L112

And by adding extra dummy buttons in `MapperParser.cpp`
https://github.com/samuelgr/Xidi/blob/dd84968cb4a97c4123a29d1f4899b0f9858a7adb/Source/MapperParser.cpp#L701-L721

I was able to get up to 31 buttons to work using custom mappers defined in xidi.ini, trying to add more buttons results in this assertion error
https://github.com/samuelgr/Xidi/blob/dd84968cb4a97c4123a29d1f4899b0f9858a7adb/Include/Xidi/Internal/ControllerTypes.h#L344-L346

If those assertions are commented out, none of the buttons work. Is there any possible way to change this?
Same happens for axes too. Other than the usual X,Y,Z,RotX,RotY,RotZ, trying to add Slider 0 and Slider 1 also results in a assertion error.

One more thing, is the possibility to add more than one POV per virtual controller too difficult to implement? I didn't even poke around to find something related to this, is just something I was thinking because DirecInput supports it.

Does something in `DataFormat.cpp` need to change to acoomodate these changes?

**Comment by samuelgr** [comment, marked as answer]
Those assertions are related to a data structure size optimization I implemented so that all of the related structures could be passed around by value very efficiently. It's a low-level optimization that most likely makes no difference whatsoever for your use case.

In files like `ControllerTypes.h` you will see some structures defined using bit-fields, like this:

https://github.com/samuelgr/Xidi/blob/dd84968cb4a97c4123a29d1f4899b0f9858a7adb/Include/Xidi/Internal/ControllerTypes.h#L204-L212

At the very end of each declaration, you will see a colon and then a number. It should be safe to remove the colon and the number, so the above struct would look like this instead:

```cpp
      struct
      {
        /// Number of axes in the virtual controller, also the number of elements of the axis type
        /// array that are valid.
        uint8_t numAxes;

        /// Number of buttons present in the virtual controller.
        uint8_t numButtons;
      };
```

That makes it safe to ignore the assertions about not fitting into some number of bits. But you'll need to do that everywhere. This can result in some assertions related to data structure size constraint violations, but those are safe to comment out after you've made this change.

As for multiple POVs, yes, that should also be possible. You would need to modify some data structures like `SState` which currently assume there is only a single POV along with the various functions that read from and write to those data structures.
